iothub/device/devdoc/API Design.md

```diff
@@ -0,0 +1,275 @@
+}
+
+public abstract class Serializer {
+    protected Serializer();
+    public abstract string ContentType { get; }
+    public abstract bool CheckWritablePropertyResponseType(object typeToCheck);
+    public abstract T ConvertFromObject<T>(object objectToConvert);
+    public abstract T DeserializeToType<T>(string stringToDeserialize);
+    public abstract string SerializeToString(object objectToSerialize);
+}
+
+public sealed class DefaultPayloadConvention : PayloadConvention {
+    public static readonly DefaultPayloadConvention Instance;
+    public DefaultPayloadConvention();
+    public override ContentEncoder PayloadEncoder { get; }
+    public override Serializer PayloadSerializer { get; }
+}
+
+public class Utf8ContentEncoder : ContentEncoder {
+    public static readonly Utf8ContentEncoder Instance;
+    public Utf8ContentEncoder();
+    public override Encoding ContentEncoding { get; }
+    public override byte[] EncodeStringToByteArray(string contentPayload);
+}
+
+public class JsonContentSerializer : Serializer {
+    public static readonly JsonContentSerializer Instance;
+    public JsonContentSerializer();
+    public override string ContentType { get; }
+    public override bool CheckWritablePropertyResponseType(object typeToCheck);
+    public override T ConvertFromObject<T>(object objectToConvert);
+    public override T DeserializeToType<T>(string stringToDeserialize);
+    public override string SerializeToString(object objectToSerialize);
+}
+
+public abstract class PayloadCollection : IEnumerable, IEnumerable<object> {
+    protected PayloadCollection(PayloadConvention payloadConvention = null);
+    public IDictionary<string, object> Collection { get; private set; }
+    public PayloadConvention Convention { get; private set; }
+    public virtual object this[string key] { get; set; }
+    public virtual void Add(string key, object value);
+    public virtual void AddOrUpdate(string key, object value);
+    public IEnumerator<object> GetEnumerator();
+    public virtual byte[] GetPayloadObjectBytes();
+    public virtual string GetSerailizedString();
+    public virtual T GetValue<T>(string key);
+    IEnumerator System.Collections.IEnumerable.GetEnumerator();
+}
+
+public static class ConventionBasedConstants {
+    public const string AckCodePropertyName = "ac";
+    public const string AckDescriptionPropertyName = "ad";
+    public const string AckVersionPropertyName = "av";
+    public const string ComponentIdentifierKey = "__t";
+    public const string ComponentIdentifierValue = "c";
+    public const string ValuePropertyName = "value";
+}
+```
+
+### Properties
+
+```csharp
+/// <summary>
+/// Retrieve the device properties.
+/// </summary>
+/// <param name="payloadConvention">A convention handler that defines the content encoding and serializer to use for commands.</param>
+/// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+/// <returns>The device properties.</returns>
```

Comment:
Add PnP here and everywhere for more context.

Reply:
We'll probably add links to the documentation surrounding the DefaultConvention implementaiton but not be explicit about PnP in our SDK and comments.

Mark and Stefan want to make sure we don't litter the SDK with PnP as that name wasn't favorable.